src/main/java/com/fasterxml/uuid/impl/TimeBasedEpochGenerator.java

```diff
@@ -0,0 +1,118 @@
+    /**********************************************************************
+     */
+
+    /**
+     * Object used for synchronizing access to timestamps, to guarantee
+     * that timestamps produced by this generator are unique and monotonically increasings.
+     * Some implementations offer even stronger guarantees, for example that
+     * same guarantee holds between instances running on different JVMs (or
+     * with native code).
+     */
+    protected final UUIDTimer _timer;
+
+
+    /**
+     * Random number generator that this generator uses.
+     */
+    protected final Random _random;
+    
+    /*
+    /**********************************************************************
+    /* Construction
+    /**********************************************************************
+     */
+    
+    /**
+     * @param rnd Random number generator to use for generating UUIDs; if null,
+     *   shared default generator is used. Note that it is strongly recommend to
+     *   use a <b>good</b> (pseudo) random number generator; for example, JDK's
+     *   {@link SecureRandom}.
+     */
+    
+    public TimeBasedEpochGenerator(UUIDTimer timer, Random rnd)
+    {
+        if (rnd == null) {
+            rnd = LazyRandom.sharedSecureRandom(); 
+        }
+        _random = rnd;
+        _timer = timer;
+    }
+    
+    /*
+    /**********************************************************************
+    /* Access to config
+    /**********************************************************************
+     */
+
+    @Override
+    public UUIDType getType() { return UUIDType.TIME_BASED_EPOCH; }
+    
+    /*
+    /**********************************************************************
+    /* UUID generation
+    /**********************************************************************
+     */
+    
+    /* As timer is not synchronized (nor _uuidBytes), need to sync; but most
+     * importantly, synchronize on timer which may also be shared between
+     * multiple instances
+     */
+    @Override
+    public UUID generate()
+    {
+        final long rawTimestamp = _timer.getTimestampV7();
+        // Time field components are kind of shuffled, need to slice:
+        int clockHi = (int) (rawTimestamp >>> 32);
+        int clockLo = (int) rawTimestamp;
+        // and dice
+        int midhi = (clockHi << 16) | (clockHi >>> 16); 
+        final byte[] buffer = new byte[10];
+        _random.nextBytes(buffer);
+        midhi = midhi | (((buffer[0] & 0xFF) << 8) + (buffer[1] & 0xFF)); 
+        // need to squeeze in type (4 MSBs in byte 6, clock hi)
+        midhi &= ~0xF000; // remove high nibble of 6th byte
+        midhi |= 0x7000; // type 7
+        long midhiL = (long) midhi;
+        midhiL = ((midhiL << 32) >>> 32); // to get rid of sign extension
+        // and reconstruct
+        long l1 = (((long) clockLo) << 32) | midhiL;
+        // last detail: must force 2 MSB to be '10'
+        long _uuidL2 = _toLong(buffer, 2); 
```

Comment:
include variant?

Reply:
That 2-bit `var` is a mystery to me... I don't get what that is, after reading and re-reading the spec. There's separate 4-bit version already so... what is this 2-bit thing. :-/

Reply:
Note to future self: "variant" is 0x10 as per:

https://www.rfc-editor.org/rfc/rfc4122#section-4.1.1

and remains the same from the earlier spec. So that's all good. Although new code does the right thing and calls `UUIDUtil.constructUUID` with `type` so it all gets set properly.